src/SDKs/Azure.Base/data-plane/Azure.Base/UriExtensions.cs

```diff
@@ -2,22 +2,174 @@
+            set
+            {
+                ResetUri();
+                if (HasQuery)
+                {
+                    _pathAndQuery.Remove(_queryIndex, _pathAndQuery.Length - _queryIndex);
+                    _queryIndex = -1;
+                }
+
+                if (!string.IsNullOrEmpty(value))
+                {
+                    _queryIndex = _pathAndQuery.Length;
+                    _pathAndQuery.Append('?');
+                    _pathAndQuery.Append(value);
+                }
+            }
+        }
+
+        public string Path
+        {
+            get => HasQuery ? _pathAndQuery.ToString(0, _queryIndex) : _pathAndQuery.ToString();
+            set
+            {
+                if (HasQuery)
+                {
+                    _pathAndQuery.Remove(0, _queryIndex);
+                    _pathAndQuery.Insert(0, value);
+                    _queryIndex = value.Length;
+                }
+                else
+                {
+                    _pathAndQuery.Remove(0, _pathAndQuery.Length);
+                    _pathAndQuery.Append(value);
+                }
+            }
+        }
+
+        private bool HasQuery => _queryIndex != -1;
+
+        public string PathAndQuery => _pathAndQuery.ToString();
+
+        public Uri Uri
         {
-            if(!string.IsNullOrEmpty(builder.Query)) {
-                builder.Query = builder.Query + "&" + name + "=" + value;
-            } 
-            else {
-                builder.Query = name + "=" + value;
+            get
+            {
+                if (_uri == null)
+                {
+                    _uri = new Uri(ToString());
+                }
+                return _uri;
+            }
+            set
+            {
+                Scheme = value.Scheme;
+                Host = value.Host;
+                Port = value.Port;
+                Path = value.AbsolutePath;
+                Query = value.Query;
+                _uri = value;
             }
         }
 
-        public static void AppendQuery(this UriBuilder builder, string name, long value)
-            => AppendQuery(builder, name, value.ToString());
+        public void AppendQuery(string name, string value)
+        {
+
+            if (!HasQuery)
+            {
+                _pathAndQuery.Append('?');
+                _queryIndex = _pathAndQuery.Length;
+            }
+            else
+            {
+                _pathAndQuery.Append('&');
+            }
+
+            _pathAndQuery.Append(name);
+            _pathAndQuery.Append('=');
+            _pathAndQuery.Append(value);
+        }
+
+        public void AppendPath(string value)
+        {
+            if (HasQuery)
+            {
+                _pathAndQuery.Insert(_queryIndex, value);
+                _queryIndex += value.Length;
+            }
+            else
+            {
+                _pathAndQuery.Append(value);
+            }
+        }
+
+        public override string ToString()
```

Comment:
Looks like the .NET [UriBuilder](https://github.com/Microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/System/net/System/uribuilder.cs) does some additional validation and formatting in its `ToString()` .   Should we consider adding some of those?

Reply:
We can have a much simplified version of `ToString` because we only need to support http style URIs. We might need to add the part about ensuring path starts with `/`